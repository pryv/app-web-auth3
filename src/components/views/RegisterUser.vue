<template>
  <div>
    <h1>Register a new user</h1>

    <v-form
      v-if="newUser==null"
      id="registerForm"
      ref="form"
      v-model="validForm"
      @submit.prevent>

      <v-text-field
        id="email"
        v-model="email"
        label="E-mail"
        placeholder="Optional (required for password reset)"
      />

      <v-text-field
        id="username"
        v-model="ctx.user.username"
        :rules="[rules.required]"
        label="Username"
      />

      <Password
        v-model="password"
        :confirmation="true"/>

      <v-autocomplete
        id="hosting"
        v-model="hosting"
        :items="hostingsSelection"
        :rules="[rules.required]"
        placeholder="Choose hosting..."
        label="Hosting"
      />

      <v-btn
        id="submitButton"
        :disabled="!validForm||submitting"
        @click="submit"
      >Create</v-btn>

      <v-btn
        id="clearButton"
        @click="clear"
      >Clear</v-btn>

      <div>
        By registering you agree with our
        <a
          target="_blank"
          href="https://pryv.com/terms-of-use/">
        terms of use</a>.
      </div>
    </v-form>
    <div v-if="ctx.isAccessRequest()">
      <v-divider class="mt-3 mb-2"/>
      <router-link :to="{ name: 'Authorization' }"><h3>Go to Sign in</h3></router-link>
    </div>

    <Alerts
      :successMsg="success"
      :errorMsg="error"/>
  </div>
</template>

<script>
import Password from './bits/Password.vue';
import Alerts from './bits/Alerts.vue';
import Context from '../../context.js';
import controllerFactory from '../controller/controller.js';

export default {
  components: {
    Password,
    Alerts,
  },
  data: () => ({
    password: '',
    email: '',
    hosting: '',
    hostingsSelection: [],
    newUser: null,
    submitting: false,
    ctx: {},
    c: null,
    error: '',
    success: '',
    rules: {
      required: value => !!value || 'This field is required.',
    },
    validForm: false,
  }),
  async created () {
    this.ctx = new Context(this.$route.query);
    await this.ctx.init();
    this.c = controllerFactory(this.ctx);
    // Fill selector with available hostings, preselect first one
    this.c.loadHostings()
      .then(this.initHostings)
      .catch(this.showError);
  },
  methods: {
    submit () {
      if (this.$refs.form.validate()) {
        this.submitting = true;

        this.generateRandomEmailIfNeeded();
        // get least occupied core in the hosting
        let availableCore = '';
        this.hostingsSelection.forEach(selection => {
          if (selection.value === this.hosting) {
            availableCore = selection.availableCore;
          }
        });
        this.c.createUser(availableCore, this.password, this.email, this.hosting)
          .then((newUser) => {
            this.newUser = newUser;
            this.success = `New user successfully created: ${newUser.username}.`;
            if (!this.ctx.isAccessRequest()) {
              location.href = this.ctx.pryvService.apiEndpointForSync(newUser.username);
            }
          })
          .catch(this.showError)
          .finally(() => { this.submitting = false; });
      }
    },
    clear () {
      this.$refs.form.reset();
    },
    initHostings (hostings) {
      this.hostingsSelection = hostings;
      if (this.hostingsSelection.length > 0) {
        this.hosting = this.hostingsSelection[0].value;
      }
    },
    showError (error) {
      this.error = error.msg;
    },
    generateRandomEmailIfNeeded () {
      if (this.email == null || (typeof this.email === 'string' && this.email.length === 0)) {
        this.email = randomUsername(20) + '@pryv.io';
      }
    },
  },
};

function randomUsername (length) {
  let result = '';
  const characters = 'abcdefghijklmnopqrstuvwxyz0123456789';
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}
</script>
