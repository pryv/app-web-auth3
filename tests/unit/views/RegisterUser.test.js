import { shallowMount } from '@vue/test-utils';
import RegisterUser from '@/components/views/RegisterUser';
import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.use(Vuetify);

describe('RegisterUser.test.js', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(RegisterUser, {
      mocks: {
        $route: {query: {}},
        $t: (key) => key,
      },
      stubs: ['router-link'],
    });
  });

  it('renders correctly (snapshots matching)', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('deactivates submit button when form is not yet filled', () => {
    const submitButton = wrapper.find('#submitButton');
    expect(wrapper.vm.validForm).toBe(false);
    expect(submitButton.attributes().disabled).toBe('true');
  });

  it('activates submit button when form is valid', () => {
    wrapper.setData({
      validForm: true,
    });
    const submitButton = wrapper.find('#submitButton');
    expect(submitButton.attributes().disabled).toBeFalsy();
  });

  it('validates email correctly', () => {
    const emailRule = wrapper.vm.rules.email;
    expect(emailRule('invalidEmail')).toBe('E-mail address is invalid');
    expect(emailRule('validEmail@test.com')).toBe(true);
  });

  it('validates required fields correctly', () => {
    const requiredRule = wrapper.vm.rules.required;
    expect(requiredRule()).toBe('This field is required');
    expect(requiredRule('notEmpty')).toBe(true);
  });
});
