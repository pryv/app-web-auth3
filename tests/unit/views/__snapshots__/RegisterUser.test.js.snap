// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterUser.test.js renders correctly (snapshots matching) 1`] = `
<div>
  <h1>
    Register a new user
  </h1>
   
  <v-form-stub
    id="registerForm"
  >
    <v-text-field-stub
      id="email"
      label="E-mail"
      rules="function required(value) {
          return !!value || 'This field is required.';
        },function email(value) {
          return (/.+@.+/.test(value) || 'E-mail must be valid.'
          );
        }"
      value=""
    />
     
    <v-text-field-stub
      id="username"
      label="Username"
      rules="function required(value) {
          return !!value || 'This field is required.';
        }"
      value=""
    />
     
    <password-stub
      confirmation="true"
      value=""
    />
     
    <v-autocomplete-stub
      allowoverflow="true"
      appendicon="$vuetify.icons.dropdown"
      backgroundcolor=""
      browserautocomplete="off"
      clearicon="$vuetify.icons.clear"
      color="primary"
      errorcount="1"
      errormessages=""
      filter="function _default(item, queryText, itemText) {
                return itemText.toLocaleLowerCase().indexOf(queryText.toLocaleLowerCase()) > -1;
            }"
      id="hosting"
      itemavatar="avatar"
      itemdisabled="disabled"
      items=""
      itemtext="text"
      itemvalue="value"
      label="Hosting"
      menuprops="[object Object]"
      messages=""
      nodatatext="$vuetify.noDataText"
      placeholder="Choose hosting..."
      rules="function required(value) {
          return !!value || 'This field is required.';
        }"
      successmessages=""
      type="text"
      value=""
      valuecomparator="function deepEqual(a, b) {
    if (a === b) return true;
    if (a instanceof Date && b instanceof Date) {
        // If the values are Date, they were convert to timestamp with getTime and compare it
        if (a.getTime() !== b.getTime()) return false;
    }
    if (a !== Object(a) || b !== Object(b)) {
        // If the values aren't objects, they were already checked for equality
        return false;
    }
    var props = Object.keys(a);
    if (props.length !== Object.keys(b).length) {
        // Different number of props, don't bother to check
        return false;
    }
    return props.every(function (p) {
        return deepEqual(a[p], b[p]);
    });
}"
    />
     
    <v-btn-stub
      activeclass="v-btn--active"
      disabled="true"
      id="submitButton"
      tag="button"
      type="button"
    >
      Create
    </v-btn-stub>
     
    <v-btn-stub
      activeclass="v-btn--active"
      id="clearButton"
      tag="button"
      type="button"
    >
      Clear
    </v-btn-stub>
     
    <div>
      
      By registering you agree with our
      
      <a
        href="https://pryv.com/terms-of-use/"
        target="_blank"
      >
        
      terms of use
      </a>
      .
    
    </div>
  </v-form-stub>
   
  <!---->
   
  <alerts-stub
    errormsg=""
    successmsg=""
  />
</div>
`;
